### DISCUSSION TITLE
Multiple scripts in a Node

### DISCUSSION THREAD

**Discussion by Tormentous** [godotengine/godot-proposals#10211] [Editor]
I have a script that is 1000 lines long, and I would really like to just have separate scripts to be much cleaner instead of having to scroll to far to find a variable, like Idk why godot is like this

**Comment by Tormentous** [comment]
Like when you click on the script button on a node maybe have a dropdown to show like all the scripts that the node uses?

**Comment by AThousandShips** [comment]
> like Idk why godot is like this

Attaching multiple scripts is enormously more difficult code wise

Related, for making dividing up code and functionality, see:
* https://github.com/godotengine/godot-proposals/issues/6416

You can also extend scripts to divide up code, but you might also need to re-evaluate your code if your scripts grow that large, for example identify code duplication, for example put the same code being used multiple times in a new function, or in some helper class, etc.

**Reply by AThousandShips** [reply]
I'd also suggest trying to use [regions](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_basics.html#code-regions) to make your code easier to navigate

**Comment by Meorge** [comment]
As someone who came to Godot from Unity and its component system, I think I understand what you're getting at. My understanding has come to be that the "one script per node" concept is part of Godot's core architecture and design philosophy. I think the "Godot-ey" equivalent of what you're looking for would be to create child nodes that contain the sub-behaviors you're thinking of. The parent can handle coordination between each of its child nodes when necessary.

**Comment by qt911025** [comment]
Separation into 2 scripts is enough: tool script and runtime script. Tool script for editor and runtime script for game, instead of @tool mode and runtime is_editor_hint assertion.
Engine itself uses macro to distinguish that, while the script uses tool annotation for all code and a simple runtime assertion, which makes no sense.

**Reply by Meorge** [reply]
I do like the idea of being able to attach an additional editor-only/tool script, although I imagine it would run into the same issue that AThousandShips suggested would exist for multiple scripts in general.

**Comment by Meorge** [comment]
This got me wondering about the possibility of scripts being able to specify/require a certain type of parent or owner node. For example, in a current project of mine, I have a state machine node with state children. Currently, I do something like `get_parent().transition_to("NextState")` to transition it to the next state. However, the type of the parent from `get_parent()` is unknown beyond just being a Node. I think it could be neat to restrict the possible parent types, so that the result of `get_parent()` is known to be a more specific type.

Something like:
```gdscript
# state_machine.gd
class_name StateMachine
extends Node
```

```gdscript
# state.gd
class_name State
extends Node
child_of StateMachine

func _ready():
    get_parent() # <-- guaranteed to be a StateMachine, autocomplete/type checking verified
```

That being said, I dunno how difficult it would be to implement, and/or if the benefits of it would outweigh the development difficulties. Just thought it'd be something to bring up as an idea, in the spirit of this conversation :)

**Comment by Tormentous** [comment]
I used different nodes with names like Attack System or Movement System or Camera System and have a script accessing variables from the player and it kinda works but It's still kind of annoying
